File: web/src/lib/api/investigations.ts
/**
 * Client-side API helpers for investigation management endpoints.
 * These functions call the SvelteKit server proxy routes which in turn
 * forward requests to the nettap-storage-daemon.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface InvestigationNote {
	id: string;
	content: string;
	created_at: string;
	updated_at: string;
}

export interface Investigation {
	id: string;
	title: string;
	description: string;
	status: string; // 'open' | 'in_progress' | 'resolved' | 'closed'
	severity: string; // 'low' | 'medium' | 'high' | 'critical'
	created_at: string;
	updated_at: string;
	alert_ids: string[];
	device_ips: string[];
	notes: InvestigationNote[];
	tags: string[];
}

export interface InvestigationStats {
	total: number;
	by_status: Record<string, number>;
	by_severity: Record<string, number>;
}

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

function buildQuery(params: Record<string, string | number | undefined>): string {
	const qs = new URLSearchParams();
	for (const [key, value] of Object.entries(params)) {
		if (value !== undefined && value !== '') {
			qs.set(key, String(value));
		}
	}
	const str = qs.toString();
	return str ? `?${str}` : '';
}

// ---------------------------------------------------------------------------
// Fetch helpers — Investigation CRUD
// ---------------------------------------------------------------------------

/**
 * Get a filtered list of investigations.
 */
export async function getInvestigations(
	opts?: { status?: string; severity?: string }
): Promise<{ investigations: Investigation[] }> {
	const query = buildQuery({ status: opts?.status, severity: opts?.severity });
	const res = await fetch(`/api/investigations${query}`);

	if (!res.ok) {
		return { investigations: [] };
	}

	return res.json();
}

/**
 * Create a new investigation.
 */
export async function createInvestigation(data: {
	title: string;
	description?: string;
	severity?: string;
	alert_ids?: string[];
	device_ips?: string[];
	tags?: string[];
}): Promise<Investigation> {
	const res = await fetch('/api/investigations', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data),
	});

	if (!res.ok) {
		return {
			id: '',
			title: data.title,
			description: data.description || '',
			status: 'open',
			severity: data.severity || 'medium',
			created_at: '',
			updated_at: '',
			alert_ids: data.alert_ids || [],
			device_ips: data.device_ips || [],
			notes: [],
			tags: data.tags || [],
		};
	}

	return res.json();
}

/**
 * Get a single investigation by ID.
 */
export async function getInvestigation(id: string): Promise<Investigation | null> {
	const res = await fetch(`/api/investigations/${encodeURIComponent(id)}`);

	if (!res.ok) {
		return null;
	}

	return res.json();
}

/**
 * Update an existing investigation.
 */
export async function updateInvestigation(
	id: string,
	data: Partial<Investigation>
): Promise<Investigation | null> {
	const res = await fetch(`/api/investigations/${encodeURIComponent(id)}`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data),
	});

	if (!res.ok) {
		return null;
	}

	return res.json();
}

/**
 * Delete an investigation.
 */
export async function deleteInvestigation(id: string): Promise<boolean> {
	const res = await fetch(`/api/investigations/${encodeURIComponent(id)}`, {
		method: 'DELETE',
	});

	return res.ok;
}

// ---------------------------------------------------------------------------
// Fetch helpers — Notes
// ---------------------------------------------------------------------------

/**
 * Add a note to an investigation.
 */
export async function addNote(
	investigationId: string,
	content: string
): Promise<InvestigationNote | null> {
	const res = await fetch(
		`/api/investigations/${encodeURIComponent(investigationId)}/notes`,
		{
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify({ content }),
		}
	);

	if (!res.ok) {
		return null;
	}

	return res.json();
}

/**
 * Update a note on an investigation.
 */
export async function updateNote(
	investigationId: string,
	noteId: string,
	content: string
): Promise<InvestigationNote | null> {
	const res = await fetch(
		`/api/investigations/${encodeURIComponent(investigationId)}/notes/${encodeURIComponent(noteId)}`,
		{
			method: 'PUT',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify({ content }),
		}
	);

	if (!res.ok) {
		return null;
	}

	return res.json();
}

/**
 * Delete a note from an investigation.
 */
export async function deleteNote(
	investigationId: string,
	noteId: string
): Promise<boolean> {
	const res = await fetch(
		`/api/investigations/${encodeURIComponent(investigationId)}/notes/${encodeURIComponent(noteId)}`,
		{
			method: 'DELETE',
		}
	);

	return res.ok;
}

// ---------------------------------------------------------------------------
// Fetch helpers — Alert linking
// ---------------------------------------------------------------------------

/**
 * Link an alert to an investigation.
 */
export async function linkAlert(
	investigationId: string,
	alertId: string
): Promise<boolean> {
	const res = await fetch(
		`/api/investigations/${encodeURIComponent(investigationId)}/alerts`,
		{
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify({ alert_id: alertId }),
		}
	);

	return res.ok;
}

/**
 * Unlink an alert from an investigation.
 */
export async function unlinkAlert(
	investigationId: string,
	alertId: string
): Promise<boolean> {
	const res = await fetch(
		`/api/investigations/${encodeURIComponent(investigationId)}/alerts/${encodeURIComponent(alertId)}`,
		{
			method: 'DELETE',
		}
	);

	return res.ok;
}

// ---------------------------------------------------------------------------
// Fetch helpers — Device linking
// ---------------------------------------------------------------------------

/**
 * Link a device IP to an investigation.
 */
export async function linkDevice(
	investigationId: string,
	deviceIp: string
): Promise<boolean> {
	const res = await fetch(
		`/api/investigations/${encodeURIComponent(investigationId)}/devices`,
		{
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify({ device_ip: deviceIp }),
		}
	);

	return res.ok;
}

// ---------------------------------------------------------------------------
// Fetch helpers — Stats
// ---------------------------------------------------------------------------

/**
 * Get investigation statistics.
 */
export async function getInvestigationStats(): Promise<InvestigationStats> {
	const res = await fetch('/api/investigations/stats');

	if (!res.ok) {
		return {
			total: 0,
			by_status: {},
			by_severity: {},
		};
	}

	return res.json();
}

File: daemon/api/investigations.py
"""
NetTap Investigation Bookmarks & Notes API Routes

Registers REST endpoints for managing investigation bookmarks, notes,
alert linking, and device association. Full CRUD with filtering and stats.
"""

import logging

from aiohttp import web

from services.investigation_store import InvestigationStore

logger = logging.getLogger("nettap.api.investigations")


# ---------------------------------------------------------------------------
# Route handlers
# ---------------------------------------------------------------------------

async def handle_list_investigations(request: web.Request) -> web.Response:
    """GET /api/investigations?status=&severity=

    List all investigations with optional status and severity filters.
    """
    store: InvestigationStore = request.app["investigation_store"]
    status = request.query.get("status")
    severity = request.query.get("severity")

    # Validate filters if provided
    if status and status not in InvestigationStore.VALID_STATUSES:
        return web.json_response(
            {"error": f"Invalid status: {status}. Must be one of {InvestigationStore.VALID_STATUSES}"},
            status=400,
        )
    if severity and severity not in InvestigationStore.VALID_SEVERITIES:
        return web.json_response(
            {"error": f"Invalid severity: {severity}. Must be one of {InvestigationStore.VALID_SEVERITIES}"},
            status=400,
        )

    investigations = store.list_all(status=status, severity=severity)
    return web.json_response({
        "investigations": [inv.to_dict() for inv in investigations],
        "count": len(investigations),
    })


async def handle_create_investigation(request: web.Request) -> web.Response:
    """POST /api/investigations

    Create a new investigation. Body: {title, description?, severity?, alert_ids?, device_ips?, tags?}
    """
    store: InvestigationStore = request.app["investigation_store"]

    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "Invalid JSON body"}, status=400)

    title = body.get("title", "").strip()
    if not title:
        return web.json_response({"error": "title is required"}, status=400)

    try:
        inv = store.create(
            title=title,
            description=body.get("description", ""),
            severity=body.get("severity", "medium"),
            alert_ids=body.get("alert_ids"),
            device_ips=body.get("device_ips"),
            tags=body.get("tags"),
        )
    except ValueError as exc:
        return web.json_response({"error": str(exc)}, status=400)

    return web.json_response(inv.to_dict(), status=201)


async def handle_get_investigation(request: web.Request) -> web.Response:
    """GET /api/investigations/{id}

    Return a single investigation by ID.
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]

    inv = store.get(inv_id)
    if inv is None:
        return web.json_response({"error": "Investigation not found"}, status=404)

    return web.json_response(inv.to_dict())


async def handle_update_investigation(request: web.Request) -> web.Response:
    """PUT /api/investigations/{id}

    Update an investigation. Body may include: title, description, status, severity, tags.
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]

    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "Invalid JSON body"}, status=400)

    try:
        inv = store.update(inv_id, **body)
    except ValueError as exc:
        return web.json_response({"error": str(exc)}, status=400)

    if inv is None:
        return web.json_response({"error": "Investigation not found"}, status=404)

    return web.json_response(inv.to_dict())


async def handle_delete_investigation(request: web.Request) -> web.Response:
    """DELETE /api/investigations/{id}

    Delete an investigation by ID.
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]

    if store.delete(inv_id):
        return web.json_response({"result": "deleted", "id": inv_id})
    return web.json_response({"error": "Investigation not found"}, status=404)


async def handle_add_note(request: web.Request) -> web.Response:
    """POST /api/investigations/{id}/notes

    Add a note to an investigation. Body: {content}
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]

    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "Invalid JSON body"}, status=400)

    content = body.get("content", "").strip()
    if not content:
        return web.json_response({"error": "content is required"}, status=400)

    note = store.add_note(inv_id, content)
    if note is None:
        return web.json_response({"error": "Investigation not found"}, status=404)

    return web.json_response(note.to_dict(), status=201)


async def handle_update_note(request: web.Request) -> web.Response:
    """PUT /api/investigations/{id}/notes/{note_id}

    Update an existing note. Body: {content}
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]
    note_id = request.match_info["note_id"]

    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "Invalid JSON body"}, status=400)

    content = body.get("content", "").strip()
    if not content:
        return web.json_response({"error": "content is required"}, status=400)

    note = store.update_note(inv_id, note_id, content)
    if note is None:
        return web.json_response({"error": "Note or investigation not found"}, status=404)

    return web.json_response(note.to_dict())


async def handle_delete_note(request: web.Request) -> web.Response:
    """DELETE /api/investigations/{id}/notes/{note_id}

    Delete a note from an investigation.
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]
    note_id = request.match_info["note_id"]

    if store.delete_note(inv_id, note_id):
        return web.json_response({"result": "deleted", "note_id": note_id})
    return web.json_response({"error": "Note or investigation not found"}, status=404)


async def handle_link_alert(request: web.Request) -> web.Response:
    """POST /api/investigations/{id}/alerts

    Link an alert to an investigation. Body: {alert_id}
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]

    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "Invalid JSON body"}, status=400)

    alert_id = body.get("alert_id", "").strip()
    if not alert_id:
        return web.json_response({"error": "alert_id is required"}, status=400)

    if store.link_alert(inv_id, alert_id):
        return web.json_response({"result": "linked", "alert_id": alert_id})
    return web.json_response({"error": "Investigation not found"}, status=404)


async def handle_unlink_alert(request: web.Request) -> web.Response:
    """DELETE /api/investigations/{id}/alerts/{alert_id}

    Unlink an alert from an investigation.
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]
    alert_id = request.match_info["alert_id"]

    if store.unlink_alert(inv_id, alert_id):
        return web.json_response({"result": "unlinked", "alert_id": alert_id})
    return web.json_response({"error": "Alert or investigation not found"}, status=404)


async def handle_link_device(request: web.Request) -> web.Response:
    """POST /api/investigations/{id}/devices

    Link a device IP to an investigation. Body: {device_ip}
    """
    store: InvestigationStore = request.app["investigation_store"]
    inv_id = request.match_info["id"]

    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "Invalid JSON body"}, status=400)

    device_ip = body.get("device_ip", "").strip()
    if not device_ip:
        return web.json_response({"error": "device_ip is required"}, status=400)

    if store.link_device(inv_id, device_ip):
        return web.json_response({"result": "linked", "device_ip": device_ip})
    return web.json_response({"error": "Investigation not found"}, status=404)


async def handle_investigation_stats(request: web.Request) -> web.Response:
    """GET /api/investigations/stats

    Return investigation statistics (counts by status and severity).
    """
    store: InvestigationStore = request.app["investigation_store"]
    return web.json_response(store.get_stats())


# ---------------------------------------------------------------------------
# Route registration
# ---------------------------------------------------------------------------

def register_investigation_routes(
    app: web.Application, store: InvestigationStore
) -> None:
    """Register all investigation API routes on the given aiohttp application."""
    app["investigation_store"] = store

    # Stats must be registered before {id} to avoid route conflict
    app.router.add_get("/api/investigations/stats", handle_investigation_stats)

    app.router.add_get("/api/investigations", handle_list_investigations)
    app.router.add_post("/api/investigations", handle_create_investigation)
    app.router.add_get("/api/investigations/{id}", handle_get_investigation)
    app.router.add_put("/api/investigations/{id}", handle_update_investigation)
    app.router.add_delete("/api/investigations/{id}", handle_delete_investigation)
    app.router.add_post("/api/investigations/{id}/notes", handle_add_note)
    app.router.add_put("/api/investigations/{id}/notes/{note_id}", handle_update_note)
    app.router.add_delete("/api/investigations/{id}/notes/{note_id}", handle_delete_note)
    app.router.add_post("/api/investigations/{id}/alerts", handle_link_alert)
    app.router.add_delete("/api/investigations/{id}/alerts/{alert_id}", handle_unlink_alert)
    app.router.add_post("/api/investigations/{id}/devices", handle_link_device)
    logger.info("Investigation API routes registered (12 endpoints)")
